src/lib.rs

```diff
@@ -186,7 +186,7 @@ pub extern "C" fn handle(
     Buffer::from_vec(v)
 }
 
-fn do_handle(
+unsafe fn do_handle(
```

Comment:
>unless they are exposed in a way that a rust caller needs to be aware of proper calling conventions not enforced by the compiler.
They are. these functions accept instances of `Buffer` which are not guaranteed to be valid when calling e.g. `read`. In this case, the function calls `Buffer::read` without checking the pointer is valid, which may be a violation of memory safety unless the caller ensured the `Buffer` is valid.

Reply:
But the only caller is an `extern "C"` function, right? And we cannot force this "unsafe" marker onto the Go calling code, can we?

If we exported `do_handle` or used it outside of `extern "C"`, I would agree with you.

In any case, if you really want to mark all these unsafe (they are, but I find this not very useful in the current context), please add Safety sections for clippy.

Reply:
>And we cannot force this "unsafe" marker onto the Go calling code, can we?

Unfortunately not

As discussed outside of GH, these functions are not unsafe at all, because they perform null checks. I removed these annotations.

Reply:
You can pass an invalid pointer to them to memory out of your control, and then get a seg fault. But there is no way to defend against this over FFI

Reply:
Yes, that is unfortunately always the case. best we can do is check against the null pointer, which is the most common mistake. We have to trust the other side to pass us pointers to valid memory.